In [1]:
# import the libraries needed
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

## Preparing the train and test dataset
---
- We have obtained the dataset from https://www.openpowerlifting.org/.
- For ease of data preprocessing, we have only taken data from the 'SBD' event where contestants are required to do all three exercises, namely Squat, Bench, Deadlift.
- From there, we randomised the data rows in the original CSV file and took out a total of 4,000 data points to be used for our train and test dataset. This dataset was further split into 3,000 for train and 1,000 for test.

In [2]:
train_test_combined = pd.read_csv('../datasets/train_test_combined.csv')
train_test_combined.head()

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
0       Alana Carrasco   F   SBD         Raw  29.5    24-34          24-39   
1       Even Dysjaland   M   SBD         Raw  19.5    20-23          19-23   
2  Alessandro Favorito   M   SBD  Single-ply  39.0    35-39          24-39   
3      Sydney Martinez   F   SBD  Single-ply   NaN      NaN            NaN   
4          Navy Villar   F   SBD  Single-ply   NaN      NaN            NaN   

        Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
0           FR-O         68.60            72  ...     Yes      USA     AZ   
1  Juniors 19-23         94.65           105  ...     Yes      NaN    NaN   
2           Open         89.40            90  ...     Yes    Italy    NaN   
3          Girls         55.16            56  ...     Yes      USA    NaN   
4          Girls         55.61            56  ...     Yes      NaN    NaN   

   Federation  ParentFederation       Date  MeetCountry  MeetState  \
0       USAPL               IPF  27/8/2017          USA         AZ   
1         NSF               IPF  13/6/2015       Norway        NaN   
2        FIPL               IPF  6/11/2004        Italy        NaN   
3      THSWPA               NaN  25/1/2014          USA         TX   
4      THSWPA               NaN  18/1/2018          USA         TX   

            MeetTown                      MeetName  
0                NaN             Sun Devil Classic  
1            Ganddal                   Klubbstevne  
2  Marina Di Carrara        Coppa Italia A Squadre  
3        JUNIOR HIGH  CARRIZO SPRINGS INVITATIONAL  
4             Sadler            S&S Lady Rams Meet  

[5 rows x 41 columns]

In [3]:
# Split combined dataset into train and test
from sklearn.model_selection import train_test_split

train,test = train_test_split(train_test_combined, test_size = 0.25, random_state = 0)
train.to_csv('../datasets/train.csv', index = False)
test.to_csv('../datasets/test.csv', index = False)

print(train.shape)
print(test.shape)

(3000, 41)
(1000, 41)


In [4]:
train_df = pd.read_csv('../datasets/train.csv')

# Removing disqualified lifters
---

Information in the `Place` column allows us to know if certain lifters were disqualified from the competition.

Values are as follows:

- Positive number: the place the lifter came in.
- G: Guest lifter. The lifter succeeded, but wasn't eligible for awards.
- DQ: Disqualified. Note that DQ could be for procedural reasons, not just failed attempts.
- DD: Doping Disqualification. The lifter failed a drug test.
- NS: No-Show. The lifter did not show up on the meet day.


We would want to remove disqualified lifters, as they may negatively impact the outcome of the predictions.

In [5]:
train_df['Place'].unique()

array(['1', '3', 'DQ', '7', '9', '2', '4', '6', 'NS', '5', '14', '11',
       '10', '8', '13', '12', '18', '15', '22', '24', '32', '26', '17',
       '19', '16', 'G', '28', 'DD', '23', '51', '21', '84', '20', '29'],
      dtype=object)

In [6]:
df = train_df.drop(train_df[
    (train_df.Place == 'DQ') |
    (train_df.Place == 'NS') |
    (train_df.Place == 'DD')
].index)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2814 non-null   object 
 1   Sex               2814 non-null   object 
 2   Event             2814 non-null   object 
 3   Equipment         2814 non-null   object 
 4   Age               1613 non-null   float64
 5   AgeClass          1884 non-null   object 
 6   BirthYearClass    1721 non-null   object 
 7   Division          2812 non-null   object 
 8   BodyweightKg      2775 non-null   float64
 9   WeightClassKg     2788 non-null   object 
 10  Squat1Kg          1284 non-null   float64
 11  Squat2Kg          1270 non-null   float64
 12  Squat3Kg          1230 non-null   float64
 13  Squat4Kg          11 non-null     float64
 14  Best3SquatKg      2787 non-null   float64
 15  Bench1Kg          1285 non-null   float64
 16  Bench2Kg          1277 non-null   float64


After dropping disqualified and no-show lifters from the dataset, we are left with only 2814 values from the original 3000 values in the train dataset. 

# Filling Missing `Age` Values
---

## Using `AgeClass`

First, use the `AgeClass` column to fill in missing values for `Age` column. 

There are a variety of possible AgeClasses that a lifter can fall into, described in https://gitlab.com/openpowerlifting/opl-data/blob/main/modules/opltypes/src/ageclass.rs 

There is an AgeClass named 5-Dec, which does not provide much information about the ages of the lifters. 

However, the `5-Dec` class has only one lifter with a missing `Age` value. We can use the median age of those in the `5-Dec` class to fill this one missing value.

In [8]:
df['AgeClass'].unique()

array(['60-64', nan, '20-23', '55-59', '13-15', '16-17', '24-34', '45-49',
       '70-74', '18-19', '40-44', '35-39', '50-54', '5-Dec', '65-69',
       '80-999', '75-79'], dtype=object)

In [9]:
## 5-Dec AgeClass, with only one missing Age Value
df[df.AgeClass == '5-Dec']

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
307           Davis Lee   M   SBD         Raw  11.0    5-Dec            NaN   
327        Brady Dibble   M   SBD         Raw   9.5    5-Dec            NaN   
350     Addyson Navarro   F   SBD         Raw  10.0    5-Dec            NaN   
411     Casey Walker #1   M   SBD         Raw  10.0    5-Dec            NaN   
1499        T. McIntosh   M   SBD  Single-ply   NaN    5-Dec            NaN   
1608            Ferrell   M   SBD       Wraps  11.5    5-Dec            NaN   
1683       Brian Foster   M   SBD  Single-ply   4.5    5-Dec            NaN   
1928  Hannah Johnson #3   F   SBD         Raw  11.5    5-Dec            NaN   
2215      Rayder Knaack   M   SBD         Raw  10.5    5-Dec            NaN   

              Division  BodyweightKg WeightClassKg  ...  Tested  Country  \
307        Youth 10-11         65.54          67.5  ...     Yes      USA   
327              MR-Y2         36.60            40  ...     Yes      USA   
350              FR-Y2         34.80            35  ...     Yes      USA   
411   Youth 11 & Under         41.96            48  ...     Yes      USA   
1499              Open        105.00           105  ...     NaN      NaN   
1608            Nov-13         82.50          82.5  ...     Yes      NaN   
1683              Open         90.00            90  ...     Yes      NaN   
1928             FR-Y3         57.56            63  ...     Yes      USA   
2215             MR-Y2         46.75            48  ...     Yes      USA   

      State  Federation  ParentFederation        Date  MeetCountry  MeetState  \
307      PA        ADAU               NaN  11/11/2012          USA         PA   
327      WY       USAPL               IPF    2/2/2019          USA         LA   
350      AZ       USAPL               IPF   23/9/2018          USA         AZ   
411      PA         RAW               NaN   7/12/2014          USA         PA   
1499    NaN        USPF               IPF   23/4/1983          USA         IN   
1608    NaN        WNPF               NaN   17/6/2000          USA         MD   
1683    NaN       ADFPA               NaN  20/11/1983          USA         NJ   
1928     LA       USAPL               IPF   1/12/2018          USA         LA   
2215     OH       USAPL               NaN   15/1/2022          USA         OH   

            MeetTown                                           MeetName  
307           Bigler                                    Central PA Open  
327              NaN                                    Youth Nationals  
350              NaN                             Apeman Strong Fest 2.0  
411       Clearfield                                    Central PA Open  
1499  West Lafayette                              Indiana State Natural  
1608       Baltimore  Teen Junior Submaster & Master Nationals & Ame...  
1683     West Orange                         New Jersey State Drug Free  
1928             NaN                                Shreveport Showdown  
2215             NaN                                      New Year Open  

[9 rows x 41 columns]

In [10]:
## find the median age of the 5-Dec AgeClass
df[df.AgeClass == '5-Dec']['Age'].median()

10.25

In [11]:
df.loc[df.AgeClass=='5-Dec', 'Age'] = df.loc[df.AgeClass=='5-Dec', 'Age'].fillna(10.5)
## StackOverflow magic https://stackoverflow.com/questions/49963491/fillna-by-selected-rows-in-pandas-dataframe 


In [12]:
df[df.AgeClass == '5-Dec']

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
307           Davis Lee   M   SBD         Raw  11.0    5-Dec            NaN   
327        Brady Dibble   M   SBD         Raw   9.5    5-Dec            NaN   
350     Addyson Navarro   F   SBD         Raw  10.0    5-Dec            NaN   
411     Casey Walker #1   M   SBD         Raw  10.0    5-Dec            NaN   
1499        T. McIntosh   M   SBD  Single-ply  10.5    5-Dec            NaN   
1608            Ferrell   M   SBD       Wraps  11.5    5-Dec            NaN   
1683       Brian Foster   M   SBD  Single-ply   4.5    5-Dec            NaN   
1928  Hannah Johnson #3   F   SBD         Raw  11.5    5-Dec            NaN   
2215      Rayder Knaack   M   SBD         Raw  10.5    5-Dec            NaN   

              Division  BodyweightKg WeightClassKg  ...  Tested  Country  \
307        Youth 10-11         65.54          67.5  ...     Yes      USA   
327              MR-Y2         36.60            40  ...     Yes      USA   
350              FR-Y2         34.80            35  ...     Yes      USA   
411   Youth 11 & Under         41.96            48  ...     Yes      USA   
1499              Open        105.00           105  ...     NaN      NaN   
1608            Nov-13         82.50          82.5  ...     Yes      NaN   
1683              Open         90.00            90  ...     Yes      NaN   
1928             FR-Y3         57.56            63  ...     Yes      USA   
2215             MR-Y2         46.75            48  ...     Yes      USA   

      State  Federation  ParentFederation        Date  MeetCountry  MeetState  \
307      PA        ADAU               NaN  11/11/2012          USA         PA   
327      WY       USAPL               IPF    2/2/2019          USA         LA   
350      AZ       USAPL               IPF   23/9/2018          USA         AZ   
411      PA         RAW               NaN   7/12/2014          USA         PA   
1499    NaN        USPF               IPF   23/4/1983          USA         IN   
1608    NaN        WNPF               NaN   17/6/2000          USA         MD   
1683    NaN       ADFPA               NaN  20/11/1983          USA         NJ   
1928     LA       USAPL               IPF   1/12/2018          USA         LA   
2215     OH       USAPL               NaN   15/1/2022          USA         OH   

            MeetTown                                           MeetName  
307           Bigler                                    Central PA Open  
327              NaN                                    Youth Nationals  
350              NaN                             Apeman Strong Fest 2.0  
411       Clearfield                                    Central PA Open  
1499  West Lafayette                              Indiana State Natural  
1608       Baltimore  Teen Junior Submaster & Master Nationals & Ame...  
1683     West Orange                         New Jersey State Drug Free  
1928             NaN                                Shreveport Showdown  
2215             NaN                                      New Year Open  

[9 rows x 41 columns]

The `Age` value has been successfully filled. We used a value of 10.5 as required by the definition of the `Age` feature in the documentation.

https://openpowerlifting.gitlab.io/opl-csv/bulk-csv-docs.html


Other points to note about the `AgeClass` column:

-   The largest class consists of lifters 80 years old and above, but the class is named `80-999`. 
-   The smallest class consists of lifters aged `5-12`.  

For simplicity, we will assume that the lifter's age is the midpoint of the boundary defined by the AgeClass. We will use string slicing to extract this middle age (last two characters of the `AgeClass` string). 

We will assume that the maximum age for the `80-999` class is 99 years.

In [13]:
df[df['Age'].isna() & df['AgeClass'].notna()]

Name Sex Event   Equipment  Age AgeClass BirthYearClass  \
27              B. Hill   M   SBD  Single-ply  NaN    24-34          24-39   
31       Monique Avalos   F   SBD  Single-ply  NaN    16-17            NaN   
43      Acasia Martinez   F   SBD  Single-ply  NaN    16-17            NaN   
61          Jamie Comer   F   SBD  Single-ply  NaN    18-19            NaN   
74        Brooke Daniel   F   SBD  Single-ply  NaN    18-19            NaN   
...                 ...  ..   ...         ...  ...      ...            ...   
2954     Ruby Hernandez   F   SBD  Single-ply  NaN    16-17          14-18   
2955  Kayla Anderson #3   F   SBD  Single-ply  NaN    18-19            NaN   
2977        Thomas Waal   M   SBD         Raw  NaN    24-34            NaN   
2980     Miranda Ovalle   F   SBD  Single-ply  NaN    18-19            NaN   
2981        J. Ashworth   M   SBD  Single-ply  NaN    35-39            NaN   

              Division  BodyweightKg WeightClassKg  ...  Tested  Country  \
27              Novice         56.00            56  ...     NaN      NaN   
31               Girls         55.34            56  ...     Yes      NaN   
43               Girls         55.16            56  ...     Yes      NaN   
61               Girls         85.18            90  ...     Yes      NaN   
74               Girls         51.26          51.9  ...     Yes      NaN   
...                ...           ...           ...  ...     ...      ...   
2954             Girls         95.53           100  ...     Yes      NaN   
2955             Girls         93.71           100  ...     Yes      NaN   
2977              Open         82.15            83  ...     Yes      NaN   
2980             Girls         66.98          67.3  ...     Yes      NaN   
2981  Submasters 35-39           NaN          125+  ...     NaN      NaN   

      State  Federation  ParentFederation        Date  MeetCountry  MeetState  \
27      NaN        USPF               IPF    5/4/1980          USA         TN   
31      NaN      THSWPA               NaN   20/2/2016          USA         TX   
43      NaN      THSWPA               NaN   18/2/2017          USA         TX   
61      NaN      THSWPA               NaN   17/2/2018          USA         TX   
74      NaN      THSWPA               NaN   26/2/2019          USA         TX   
...     ...         ...               ...         ...          ...        ...   
2954    NaN      THSWPA               NaN    3/3/2015          USA         TX   
2955    NaN      THSWPA               NaN   24/1/2019          USA         TX   
2977    NaN         NSF               IPF  16/12/2017       Norway        NaN   
2980    NaN      THSWPA               NaN   14/2/2019          USA         TX   
2981    NaN        USPF               IPF   18/5/1991          USA         NC   

          MeetTown                         MeetName  
27     Chattanooga       Southeastern Open & Novice  
31    Seguin Texas              Seguin Matador Meet  
43     Altair Rice  Rice Consolidated Saturday Meet  
61       Normangee      Normangee Powerlifting Meet  
74      GHS Indoor      Region 6 Div 2 Championship  
...            ...                              ...  
2954    Gatesville         Reg 2 Div 3 Championship  
2955     Henderson                Lion Invitational  
2977         Skien                     Åpent stevne  
2980       TSS Gym          TSS-UTSA Invitational 6  
2981    High Point                Masters Nationals  

[270 rows x 41 columns]

In [23]:
def newString(g):
    ## Function that returns the midpoint of an AgeClass 
    if type(g) == str:
        low = high = ""
        for x in range(len(g)):
            if g[x] == '-':
                high += g[x+1]
                high += g[x+2]
                break
            low += g[x]
        return (float(low)+float(high)) / 2
    return g

In [18]:
## First, fill the values using the AgeClass Strings
df['Age'].fillna(df.AgeClass, inplace=True)

In [19]:
## Use the newString function to replace the Strings with the maximum age 
df['Age'] = df['Age'].apply(newString)

In [22]:
## cell for checking
df.loc[2981]

Name                      J. Ashworth
Sex                                 M
Event                             SBD
Equipment                  Single-ply
Age                              37.0
AgeClass                        35-39
BirthYearClass                    NaN
Division             Submasters 35-39
BodyweightKg                      NaN
WeightClassKg                    125+
Squat1Kg                          NaN
Squat2Kg                          NaN
Squat3Kg                          NaN
Squat4Kg                          NaN
Best3SquatKg                    237.5
Bench1Kg                          NaN
Bench2Kg                          NaN
Bench3Kg                          NaN
Bench4Kg                          NaN
Best3BenchKg                    145.0
Deadlift1Kg                       NaN
Deadlift2Kg                       NaN
Deadlift3Kg                       NaN
Deadlift4Kg                       NaN
Best3DeadliftKg                 242.5
TotalKg                         625.0
Place       

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2814 non-null   object 
 1   Sex               2814 non-null   object 
 2   Event             2814 non-null   object 
 3   Equipment         2814 non-null   object 
 4   Age               1884 non-null   float64
 5   AgeClass          1884 non-null   object 
 6   BirthYearClass    1721 non-null   object 
 7   Division          2812 non-null   object 
 8   BodyweightKg      2775 non-null   float64
 9   WeightClassKg     2788 non-null   object 
 10  Squat1Kg          1284 non-null   float64
 11  Squat2Kg          1270 non-null   float64
 12  Squat3Kg          1230 non-null   float64
 13  Squat4Kg          11 non-null     float64
 14  Best3SquatKg      2787 non-null   float64
 15  Bench1Kg          1285 non-null   float64
 16  Bench2Kg          1277 non-null   float64


In [99]:
## cell for checking if Age was filled correctly
df.loc[2981]

Name                      J. Ashworth
Sex                                 M
Event                             SBD
Equipment                  Single-ply
Age                              39.0
AgeClass                        35-39
BirthYearClass                    NaN
Division             Submasters 35-39
BodyweightKg                      NaN
WeightClassKg                    125+
Squat1Kg                          NaN
Squat2Kg                          NaN
Squat3Kg                          NaN
Squat4Kg                          NaN
Best3SquatKg                    237.5
Bench1Kg                          NaN
Bench2Kg                          NaN
Bench3Kg                          NaN
Bench4Kg                          NaN
Best3BenchKg                    145.0
Deadlift1Kg                       NaN
Deadlift2Kg                       NaN
Deadlift3Kg                       NaN
Deadlift4Kg                       NaN
Best3DeadliftKg                 242.5
TotalKg                         625.0
Place       

## Using `BirthYearClass`


Now the age column only contains numeric float64 datatypes, which shows that we have successfully replaced all the strings after the fillna() operation.

However, there are still some missing values, where both `Age` and `AgeClass` data was missing.

We can fill the remaining missing values for `Age` using the `BirthYearClass` information.


Information about `BirthYearClass`: 

     The ages in the range are the oldest possible ages for the lifter that year. For example, 40-49 means "the year the lifter turns 40 through the full year in which the lifter turns 49."

     BirthYearClass is used primarily by the IPF and by IPF affiliates. Non-IPF federations tend to use AgeClass instead.

     The full range available to BirthYearClass is defined by modules/opltypes/src/birthyearclass.rs. 

We will also use the midpoint of the range defined in `BirthYearClass` to fill out the missing `Age` values. This method is similar to the approach used with `AgeClass`.

In [124]:
df[df['Age'].isna() & df['BirthYearClass'].notna()]

Name Sex Event   Equipment  Age AgeClass BirthYearClass  \
27             B. Hill   M   SBD  Single-ply  NaN    24-34          24-39   
40         Ioan Ordean   M   SBD  Single-ply  NaN      NaN          19-23   
54    John Mitchell #7   M   SBD         Raw  NaN      NaN          50-59   
137         D. Mullins   M   SBD  Single-ply  NaN    40-44          40-49   
183       Glenn France   M   SBD  Single-ply  NaN    16-17          14-18   
...                ...  ..   ...         ...  ...      ...            ...   
2867   Henry Francisco   M   SBD         Raw  NaN    20-23          19-23   
2889   Taylor Lewis #1   F   SBD  Single-ply  NaN    16-17          14-18   
2912      Nathan Crews   M   SBD       Wraps  NaN      NaN          24-39   
2954    Ruby Hernandez   F   SBD  Single-ply  NaN    16-17          14-18   
2993        Lily Riggs   F   SBD         Raw  NaN      NaN          19-23   

           Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
27           Novice         56.00            56  ...     NaN      NaN    NaN   
40         Junior I        122.50          120+  ...     Yes      NaN    NaN   
54           M-C-M2         99.50           100  ...     Yes  England     WM   
137   Masters 40-44        125.00           125  ...     NaN      NaN    NaN   
183            M-T2         65.00          67.5  ...     Yes      USA    NaN   
...             ...           ...           ...  ...     ...      ...    ...   
2867        Juniors         77.80            83  ...     Yes   Canada     SK   
2889          Girls         53.98            56  ...     Yes      NaN    NaN   
2912     Submasters         98.07           100  ...     NaN      NaN    NaN   
2954          Girls         95.53           100  ...     Yes      NaN    NaN   
2993        Juniors         69.15            72  ...     Yes   Canada     BC   

      Federation  ParentFederation        Date  MeetCountry  MeetState  \
27          USPF               IPF    5/4/1980          USA         TN   
40          FRPL               IPF    1/6/2013      Romania        NaN   
54            BP               IPF    4/9/2010      England        NaN   
137         USPF               IPF   13/5/1989          USA         TX   
183        USAPL               IPF  22/11/2003          USA         TX   
...          ...               ...         ...          ...        ...   
2867         CPU               IPF   21/5/2016       Canada         SK   
2889      THSWPA               NaN   13/2/2019          USA         TX   
2912         SPF               NaN   9/12/2017          USA         TN   
2954      THSWPA               NaN    3/3/2015          USA         TX   
2993         CPU               IPF    3/3/2020       Canada         MB   

         MeetTown                                           MeetName  
27    Chattanooga                         Southeastern Open & Novice  
40            NaN  Campionatul National de Powerlifting Juniori s...  
54     Birmingham                    British Unequipped Championship  
137        Dallas                                  Masters Nationals  
183        Austin                                      Longhorn Open  
...           ...                                                ...  
2867       Regina                                    War Of The West  
2889      Baytown                                      GCM LQM Girls  
2912    Cleveland                                    Gritmas Classic  
2954   Gatesville                           Reg 2 Div 3 Championship  
2993     Winnipeg                                          Nationals  

[156 rows x 41 columns]

In [125]:
df['BirthYearClass'].unique()

array(['60-69', nan, '50-59', '14-18', '24-39', '40-49', '70-999',
       '19-23'], dtype=object)

In [24]:
## First, fill the values using the AgeClass Strings
df['Age'].fillna(df.BirthYearClass, inplace=True)

In [25]:
## Use the newString function to replace the Strings with the maximum age 
df['Age'] = df['Age'].apply(newString)

In [26]:
## cell for checking
df.loc[2993]

Name                Lily Riggs
Sex                          F
Event                      SBD
Equipment                  Raw
Age                       21.0
AgeClass                   NaN
BirthYearClass           19-23
Division               Juniors
BodyweightKg             69.15
WeightClassKg               72
Squat1Kg                 135.0
Squat2Kg                 142.5
Squat3Kg                -145.0
Squat4Kg                   NaN
Best3SquatKg             142.5
Bench1Kg                  72.5
Bench2Kg                 -77.5
Bench3Kg                  77.5
Bench4Kg                   NaN
Best3BenchKg              77.5
Deadlift1Kg              160.0
Deadlift2Kg              165.0
Deadlift3Kg              170.0
Deadlift4Kg                NaN
Best3DeadliftKg          170.0
TotalKg                  390.0
Place                        2
Dots                    397.11
Wilks                   391.29
Glossbrenner            345.01
Goodlift                  80.7
Tested                     Yes
Country 

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2814 non-null   object 
 1   Sex               2814 non-null   object 
 2   Event             2814 non-null   object 
 3   Equipment         2814 non-null   object 
 4   Age               1941 non-null   float64
 5   AgeClass          1884 non-null   object 
 6   BirthYearClass    1721 non-null   object 
 7   Division          2812 non-null   object 
 8   BodyweightKg      2775 non-null   float64
 9   WeightClassKg     2788 non-null   object 
 10  Squat1Kg          1284 non-null   float64
 11  Squat2Kg          1270 non-null   float64
 12  Squat3Kg          1230 non-null   float64
 13  Squat4Kg          11 non-null     float64
 14  Best3SquatKg      2787 non-null   float64
 15  Bench1Kg          1285 non-null   float64
 16  Bench2Kg          1277 non-null   float64


In [34]:
df.groupby('WeightClassKg').mean()

Age  BodyweightKg    Squat1Kg    Squat2Kg    Squat3Kg  \
WeightClassKg                                                                
100            31.274074     97.748100  166.046267  135.497838   74.515634   
100+           17.500000    112.214545         NaN         NaN         NaN   
103            33.250000     99.995000         NaN         NaN         NaN   
105            30.870370    101.130612  164.765333  162.131081   63.778571   
108.2                NaN    108.200000         NaN         NaN         NaN   
...                  ...           ...         ...         ...         ...   
88                   NaN     86.050000  195.040000  199.580000  204.120000   
90             30.274510     88.286026  139.825692  102.203284   35.634262   
90+            32.333333    104.674444   97.142857   51.785714  108.571429   
93             28.266667     89.831504  131.890741  150.063889   55.040952   
99.7                 NaN     98.860000  181.435000  190.510000   10.205000   

               Squat4Kg  Best3SquatKg    Bench1Kg    Bench2Kg    Bench3Kg  \
WeightClassKg                                                               
100              -225.0    224.016784  118.674800   64.543919  -22.942222   
100+                NaN    143.070833         NaN         NaN         NaN   
103                 NaN    241.250000         NaN         NaN         NaN   
105                 NaN    218.190816  124.176000   89.671233   -1.591549   
108.2               NaN    265.000000         NaN         NaN         NaN   
...                 ...           ...         ...         ...         ...   
88                  NaN    204.120000  120.200000  124.740000 -131.540000   
90                142.5    193.542212  107.792239   74.148657  -12.120313   
90+                 NaN    149.351429   73.214286   32.142857   34.642857   
93                  NaN    206.799248  101.207407   80.565741   14.272897   
99.7                NaN    193.910000  130.405000  140.610000  -11.340000   

               ...  Deadlift1Kg  Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  \
WeightClassKg  ...                                                       
100            ...   179.222667   176.997671    44.825652       -271.0   
100+           ...          NaN          NaN          NaN          NaN   
103            ...          NaN          NaN          NaN          NaN   
105            ...   209.556000   169.160811    37.471831          NaN   
108.2          ...          NaN          NaN          NaN          NaN   
...            ...          ...          ...          ...          ...   
88             ...   224.530000   231.330000   235.870000          NaN   
90             ...   190.823731   144.695970     2.805410         97.5   
90+            ...   164.642857    51.785714   -58.214286          NaN   
93             ...   189.059259   168.349074     5.006731          NaN   
99.7           ...   218.860000   232.470000    -5.670000          NaN   

               Best3DeadliftKg     TotalKg        Dots       Wilks  \
WeightClassKg                                                        
100                 229.478090  598.372250  382.683850  377.573950   
100+                141.370000  359.095000  271.666364  267.901818   
103                 256.250000  671.250000  413.400000  408.775000   
105                 244.185567  605.963265  371.227653  367.257755   
108.2               290.000000  732.500000  436.570000  433.310000   
...                        ...         ...         ...         ...   
88                  235.870000  564.720000  373.760000  369.220000   
90                  208.726947  525.607904  362.156376  356.303974   
90+                 166.557857  400.899286  338.745556  332.595556   
93                  230.181955  576.418797  372.877820  368.189774   
99.7                237.005000  574.930000  355.630000  351.505000   

               Glossbrenner   Goodlift  
WeightClassKg                           
100              357.283500  69.820700  
100+   

In [102]:
t = df.groupby('Division')['Age'].median()
t ## Series showing median age of each division

Division
13-16      15.5
17-19      19.0
18 & 19    18.5
18-19      18.5
40-49       NaN
           ... 
wnov       31.5
wopen      32.5
wpure       NaN
wsm2       39.0
wteen      17.0
Name: Age, Length: 382, dtype: float64

In [103]:
df['Age'] = df['Age'].fillna(df.groupby('Division')['Age'].transform('median'))

##stackoverflow magic https://stackoverflow.com/questions/19966018/pandas-filling-missing-values-by-mean-in-each-group


In [106]:
## cell for checking. If the numbers are the same, the values have been filled correctly
print(df.loc[3]['Age']) ## value in the Age column after the fillna()
print(t[df.loc[3]['Division']]) ## median value of the division

17.75
17.75


Name Sex Event   Equipment  Age AgeClass BirthYearClass  \
40         Ioan Ordean   M   SBD  Single-ply  NaN      NaN          19-23   
1014        Al Statman   M   SBD  Single-ply  NaN      NaN          60-69   
1117  Adam Anderson #1   M   SBD         Raw  NaN      NaN          40-49   
1655     Jeff Dispensa   M   SBD       Wraps  NaN      NaN          50-59   
2781    Chris Baker #1   M   SBD         Raw  NaN      NaN          40-49   

      Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
40    Junior I         122.5          120+  ...     Yes      NaN    NaN   
1014      M-M3          90.0            90  ...     Yes      USA    NaN   
1117    M-C-M1         102.7           105  ...     Yes  England    NaN   
1655     50-59          90.0            90  ...     Yes      NaN    NaN   
2781    M-C-M1         140.0          120+  ...     Yes  England    YNE   

      Federation  ParentFederation        Date  MeetCountry  MeetState  \
40          FRPL               IPF    1/6/2013      Romania        NaN   
1014       USAPL               IPF   15/5/2004          USA         VA   
1117          BP               IPF  20/10/2018           UK        NaN   
1655        WNPF               NaN  14/12/2014          USA         NJ   
2781         EPA               IPF   17/9/2017      England        NaN   

        MeetTown                                           MeetName  
40           NaN  Campionatul National de Powerlifting Juniori s...  
1014      Oakton                                Nations Capital Cup  
1117     Belfast  British Classic Masters Powerlifting Champions...  
1655  Bordentown  Serge McCray Powerlifting & Single Lift Champi...  
2781         NaN                                        YNE Masters  

[5 rows x 41 columns]

From the documentation (https://openpowerlifting.gitlab.io/opl-csv/bulk-csv-docs.html): 

          Ages can be one of two types: exact or approximate. Exact ages are given as integer numbers, for example 23. Approximate ages are given as an integer plus 0.5, for example 23.5.
     
          A lifter with the given age 23.5 could be either 23 or 24 -- we don't have enough information to know.

Given that the median values we have used to fill the data could be out of the defined values for age (e.g. 17.75), it would help for us to round off all the Age values to the nearest whole number

In [107]:
df['Age'] = df['Age'].apply(lambda x: round(x, 0))

In [110]:
df['Age'].unique()

array([61., 28., 18., 23., 58., 14., 16., 25., 34., 46., 73., 21., 41.,
       30., 17., 27., nan, 22., 45., 39., 20., 54., 19., 32., 33., 38.,
       40., 36., 24., 37., 29., 44., 26., 60., 50., 31., 42., 59., 64.,
       57., 49., 15., 52., 43., 56., 11., 53., 10., 51., 55., 62., 12.,
       35., 68., 13., 65., 48., 72., 66., 80., 70., 47., 75.,  4., 69.,
       71., 74.])

In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2814 non-null   object 
 1   Sex               2814 non-null   object 
 2   Event             2814 non-null   object 
 3   Equipment         2814 non-null   object 
 4   Age               2752 non-null   float64
 5   AgeClass          1884 non-null   object 
 6   BirthYearClass    1721 non-null   object 
 7   Division          2812 non-null   object 
 8   BodyweightKg      2775 non-null   float64
 9   WeightClassKg     2788 non-null   object 
 10  Squat1Kg          1284 non-null   float64
 11  Squat2Kg          1270 non-null   float64
 12  Squat3Kg          1230 non-null   float64
 13  Squat4Kg          11 non-null     float64
 14  Best3SquatKg      2787 non-null   float64
 15  Bench1Kg          1285 non-null   float64
 16  Bench2Kg          1277 non-null   float64


Drop the remaining values NA values for age

In [ ]:
df.dropna(subset=['Age'], inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2793 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2793 non-null   object 
 1   Sex               2793 non-null   object 
 2   Event             2793 non-null   object 
 3   Equipment         2793 non-null   object 
 4   Age               2793 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2114 non-null   object 
 7   Division          2793 non-null   object 
 8   BodyweightKg      2790 non-null   float64
 9   WeightClassKg     2780 non-null   object 
 10  Squat1Kg          1074 non-null   float64
 11  Squat2Kg          1066 non-null   float64
 12  Squat3Kg          1044 non-null   float64
 13  Squat4Kg          4 non-null      float64
 14  Best3SquatKg      1850 non-null   float64
 15  Bench1Kg          1696 non-null   float64
 16  Bench2Kg          1687 non-null   float64


# Filling Missing Values (Bodyweight)

In [ ]:
df[df.BodyweightKg.isna()]

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
1086   Richard Nowazek   M   SBD  Single-ply  51.5    50-54          50-59   
1597  Madison Neesmith   F     B         Raw  20.0    20-23          19-23   
1605         Jay Bakke   M   SBD   Multi-ply  53.0    50-54          50-59   

          Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
1086          Open           NaN          140+  ...     Yes   Canada     BC   
1597            JR           NaN         82.5+  ...     Yes      USA     NC   
1605  M_MEM_3_AAPF           NaN           NaN  ...     Yes      USA    NaN   

      Federation  ParentFederation       Date  MeetCountry  MeetState  \
1086         CPU               IPF   6/4/2002       Canada         BC   
1597        WNPF               NaN  6/10/2018          USA        NaN   
1605         APF               WPC  17/7/2010          USA         MT   

       MeetTown                           MeetName  
1086  Vancouver        BC Provincial Championships  
1597        NaN  World Tournament Of Championships  
1605        NaN                Big Sky State Games  

[3 rows x 41 columns]

We can use the weightclasses to fill in the first 2 missing bodyweight values. We assume they hold the minimum weight in the particular weightclass

Weight classes can be specified as a maximum or as a minimum. Maximums are specified by just the number, for example `90` means "up to (and including) 90kg." minimums are specified by a `+` to the right of the number, for example `90+` means "above (and excluding) 90kg.

In [ ]:
df['BodyweightKg'].fillna(df.WeightClassKg, inplace=True)

In [ ]:
df['BodyweightKg'] = df['BodyweightKg'].apply(lambda f: float(f[:-1]) if type(f) == str else f)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2793 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2793 non-null   object 
 1   Sex               2793 non-null   object 
 2   Event             2793 non-null   object 
 3   Equipment         2793 non-null   object 
 4   Age               2793 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2114 non-null   object 
 7   Division          2793 non-null   object 
 8   BodyweightKg      2792 non-null   float64
 9   WeightClassKg     2780 non-null   object 
 10  Squat1Kg          1074 non-null   float64
 11  Squat2Kg          1066 non-null   float64
 12  Squat3Kg          1044 non-null   float64
 13  Squat4Kg          4 non-null      float64
 14  Best3SquatKg      1850 non-null   float64
 15  Bench1Kg          1696 non-null   float64
 16  Bench2Kg          1687 non-null   float64


The final missing bodyweight value belongs to a Male in the 50-54 AgeClass. We can use the median value of that AgeClass to fill the remaining missing value.

In [ ]:
df.groupby(['AgeClass']).median()

Age  BodyweightKg  Squat1Kg  Squat2Kg  Squat3Kg  Squat4Kg  \
AgeClass                                                                
13-15     14.50        58.700     98.75     96.50     83.75      65.0   
16-17     16.50        70.220    115.00    122.50     96.50       NaN   
18-19     18.50        73.900    137.50    147.50    112.50       NaN   
20-23     21.50        81.200    170.00    170.00    118.75     165.0   
24-34     28.00        83.000    147.50    157.50    127.50     241.0   
35-39     37.00        86.910    140.00    117.50    117.50       NaN   
40-44     42.00        91.900    165.00    158.76    117.50       NaN   
45-49     47.00        89.050    123.75    151.25    132.50       NaN   
5-Dec     11.00        39.900     30.00     45.00    -38.75       NaN   
50-54     52.00        82.450    150.00    120.00     82.50       NaN   
55-59     57.00        89.225    110.00    107.50    132.50       NaN   
60-64     61.75        81.890    120.00    135.00     80.00       NaN   
65-69     66.00        74.150     71.25     53.75    -52.50       NaN   
70-74     71.75        85.100   -100.00    105.00    107.50       NaN   
75-79     77.50        80.360     85.00     92.50     96.25       NaN   
80-999    81.50        75.000       NaN       NaN       NaN       NaN   

          Best3SquatKg  Bench1Kg  Bench2Kg  Bench3Kg  ...  Deadlift1Kg  \
AgeClass                                              ...                
13-15          104.330    52.500     57.50    37.500  ...      108.750   
16-17          126.005    70.000     62.50   -35.875  ...      128.750   
18-19          142.500    87.500     90.00   -60.000  ...      170.000   
20-23          186.250   105.000    107.50    56.250  ...      185.000   
24-34          182.500   120.000    115.00   -55.000  ...      190.000   
35-39          193.750   119.000     90.00   -52.500  ...      177.500   
40-44          190.000   132.500    117.50   -67.500  ...      200.000   
45-49          182.500   130.000    105.00   -50.000  ...      150.000   
5-Dec           25.000    23.750     20.00   -22.500  ...       42.215   
50-54          183.750   105.000     96.25   -57.500  ...      185.000   
55-59          150.000    99.895    102.75    60.000  ...      152.500   
60-64          145.000    95.000     83.75   -47.500  ...      162.500   
65-69          130.000    38.750     60.00     1.250  ...      100.000   
70-74          105.000    80.000     90.00    90.000  ...      140.000   
75-79           96.250    72.500     77.50    -3.750  ...      122.500   
80-999         153.750    97.520    102.06   104.330  ...          NaN   

          Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg  \
AgeClass                                                                    
13-15          118.75       101.05      -105.00           115.00   220.00   
16-17          138.75       130.00      -226.25           137.50   300.00   
18-19          173.75       125.00      -125.00           154.22   331.12   
20-23          190.00       130.00       327.50           212.50   422.50   
24-34          190.00       126.25       125.00           210.00   320.00   
35-39          156.25       142.50       137.50           210.00   267.50   
40-44          203.75       120.00          NaN           218.75   240.00   
45-49          160.00       157.50       188.00           192.64   225.00   
5-Dec           63.50       -40.00       -86.18            35.00    35.00   
50-54          200.00       133.27          NaN           205.00   231.33   
55-59          165.00       155.00          NaN           180.00   326.50   
60-64          180.00       130.00          NaN           185.00   182.50   
65-69          107.50       105.00        85.00           120.00   185.00   
70-74          157.00       162.00          NaN           160.00   288.75   
75-79          130.00        -3.75          NaN           130.00   110.00   
80-999            NaN          NaN          NaN        

In [ ]:
df.BodyweightKg.fillna(82.45, inplace=True)

In [1]:
print('End of file ')

End of file 
